src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -345,6 +345,86 @@ object MatrixIR {
       keepF(mv, keep)
     })
   }
+
+  def collectColsByKey(typ: MatrixType): (MatrixType, MatrixValue => MatrixValue) = {
+    val newColValueType = TStruct(typ.colValueStruct.fields.map(f => f.copy(typ = TArray(f.typ, required = true))))
+    val newColType = typ.colKeyStruct ++ newColValueType
+    val newEntryType = TStruct(typ.entryType.fields.map(f => f.copy(typ = TArray(f.typ, required = true))))
+    val newMatrixType = typ.copyParts(colType = newColType, entryType = newEntryType)
+    val oldRVRowType = typ.rvRowType
+    val newRVRowType = newMatrixType.rvRowType
+    val oldEntryArrayType = typ.entryArrayType
+    val oldEntryType = typ.entryType
+    val localRowSize = newRVRowType.size
+
+    (newMatrixType, { mv =>
+      val colValMap: Map[Row, Array[Int]] = mv.colValues.value.map(_.asInstanceOf[Row]).zipWithIndex
+        .groupBy[Row]{ case (r, i) => typ.extractColKey(r)}
+        .mapValues{ _.map{ case (r, i) => i }.toArray}
+      val idxMap = colValMap.values.toArray
+
+      val newColValues: BroadcastIndexedSeq = mv.colValues.copy(
+        value = colValMap.map { case (key, idx) =>
+          Row.fromSeq(key.toSeq ++ newColValueType.fields.map { f =>
+            idx.map { i =>
+              mv.colValues.value(i).asInstanceOf[Row]
+                .get(typ.colType.fieldIdx(f.name))
```

Comment:
Why not pull `typ.colType.fieldIdx(f.name)` out of the inner inner loop, e.g.,
```
      val fieldIdxMap = newColValueType.fields.map(f => typ.colType.fieldIdx(f.name)).toArray
      val newColValues: BroadcastIndexedSeq = mv.colValues.copy(
        value = colValMap.map { case (key, idx) =>
          Row.fromSeq(key.toSeq ++ fieldIdxMap.map { j =>
            idx.map(i => mv.colValues.value(i).asInstanceOf[Row].get(j)).toIndexedSeq
          })
        }.toIndexedSeq)
```

Reply:
Is this for readability, or for optimization? I'm not convinced this is better on either front.

Reply:
As we discussed, optimization. The code uses a hashmap of string to int with random access, which it could use the array of indices of the value fields in order.